dist/bash-completion/tpm2-tools/tpm2_completion.bash

```diff
@@ -0,0 +1,153 @@
+            suggestions=( $( compgen -W "${halg[*]}" -- "$cur" ) )
+            COMPREPLY=("${suggestions[@]}")
+          else
+            _filedir
+          fi
+          return;;
+
+      -h | --help)
+          suggestions=( $( compgen -W 'summary manpage' -- "$cur" ) )
+          COMPREPLY=("${suggestions[@]}")
+          return;;
+      -T)
+          suggestions=( $( compgen -W "${tcti_opts[*]}" -- "$cur" ) )
+          COMPREPLY=("${suggestions[@]}")
+          [[ $COMPREPLY == *: ]] && compopt -o nospace
+          return;;
+      -f)
+          if [[ "${COMP_WORDS[0]}" == "tpm2_activatecredential" || "${COMP_WORDS[0]}" == "print" ]]; then
+            _filedir
+          elif [[ "${COMP_WORDS[0]}" == "tpm2_quote" || "${COMP_WORDS[0]}" == "tpm2_sign" ]]; then
+            suggestions=( $( compgen -W 'plain tss' -- "$cur" ) )
+            COMPREPLY=("${suggestions[@]}")
+          elif [[ "${COMP_WORDS[0]}" == "tpm2_verifysignature" ]]; then
+            suggestions=( $( compgen -W "${signing_schemes[*]}" -- "$cur" ) )
+            COMPREPLY=("${suggestions[@]}")
+          fi
+          return;;
+      -s)
+          if [[ "${COMP_WORDS[0]}" == "tpm2_createak" ]]; then
+            suggestions=( $( compgen -W "${signing_alg[*]}" -- "$cur" ) )
+            COMPREPLY=("${suggestions[@]}")
+          elif [["${COMP_WORDS[0]}" == "tpm2_verifysignature" ]];then
+            _filedir
+          fi
+          return;;
+      -u | -r)
+          if [[ "${COMP_WORDS[0]}" == "tpm2_load" || "${COMP_WORDS[0]}" == "tpm2_loadexternal" ]]; then
+            _filedir
+          fi
+          return;;
+      -I)
+          if [[ "${COMP_WORDS[0]}" != "tpm2_nvdefine" ]]; then
+            _filedir
+          fi
+          return;;
+      -k | -K)
+          if [[ "${COMP_WORDS[0]}" == "tpm2_import" ]]; then
+            _filedir
+          fi
+          return;;
+      -i)
+          if [[ "${COMP_WORDS[0]}" == "tpm2_send" ]]; then
+            _filedir
+          fi
+          return;;
+      -m)
+          if [[ "${COMP_WORDS[0]}" != "tpm2_quote" ]]; then
+            _filedir
+          fi
+          return;;
+      -t)
+          if [[ "${COMP_WORDS[0]}" == "tpm2_sign" ]]; then
+            _filedir
+          fi
+          return;;
+      -L)
+          if [[ "${COMP_WORDS[0]}" == "tpm2_create" || "${COMP_WORDS[0]}" == "tpm2_createprimary" || "${COMP_WORDS[0]}" == "tpm2_nvdefine" ]]; then
+            _filedir
+          fi
+          return;;
+      -e)
+          if [[ "${COMP_WORDS[0]}" == "tpm2_makecredential" ]]; then
+            _filedir
+          fi
+          return;;
+
+      -S | F | C)
+         _filedir
+         return;;
+
+    esac
+
+    if [[ "$cur" == -* ]]; then #start completion
+        _exclude_completed_opts
+        COMPREPLY=( $( compgen -W '$( echo ${suggestions[@]//<value>} )' -- "$cur" ) )
+        [[ $COMPREPLY == *= ]] && compopt -o nospace
+        return
+    fi
```

Comment:
I would also add here:

```
 COMPREPLY=( $( compgen -W '$( echo ${suggestions[@]//<value>} )' -- "$cur" ) )
```

Because when I tested, I typed `tpm2_create` then a space and then pressed a tab but nothing happened. I had to look at your code to understand that I first had to type `-` for the options to be listed. That makes sense for someone who knows the tool but someone using for the first type wouldn't know if there's a subcommand or just options.

Adding the mentioned line would complete a `-` for the user.